### ISSUE TITLE
'invalid namespace for attributes' error when adding custom namespaces to svg element

### ISSUE THREAD

**Issue by gopeter** [mbostock/d3#1935]
When trying to add a custom namespace as attribute to the `svg` element, I get the following error: `Uncaught NamespaceError: Failed to execute 'setAttributeNS' on 'Element': 'http://www.w3.org/2000/xmlns/' is an invalid namespace for attributes. `

I need to add custom namespaces to the `svg` element to create a valid SVG document that can be used to extract semantic data (e.g. [with the W3C RDFa validator](http://www.w3.org/2012/pyRdfa/Validator.html#distill_by_input))

Thats how I tried it:
````
// register custom namespaces
d3.ns.prefix.ex = 'http://example.com/';

// create chart
var chart = d3.select('#chart')
            .append('svg:svg')
            .attr('width', width)
            .attr('version', '1.1')
            .attr('xmlns', 'http://www.w3.org/2000/svg')
            .attr('xmlns:ex', 'http://example.com/');
````

The result should looks like this:

`<svg version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:ex="http://example.com">`

**Comment by mbostock** [comment]
So, instead of:

```js
element.setAttributeNS("http://www.w3.org/2000/xmlns/", "ex", "http://example.com/");
```

We should be saying:

```js
element.setAttributeNS("http://www.w3.org/2000/xmlns/", "xmlns:ex", "http://example.com/");
```

Is that it?

**Comment by jasondavies** [comment]
Correct!

**Comment by hugolpz** [comment]
Is this implemented ?

**Comment by hugolpz** [comment]
Implemented ?

**Comment by micahstubbs** [comment]
_"By the way, for your specific example, you can set the xmlns attributes using the DOM API directly by using selection.node() to retrieve the first Element in the selection."_

I'll share the code I used for workaround that @jasondavies describes:
        
    var svg = d3.select("body").append("svg")

    svg.node().setAttributeNS("http://www.w3.org/2000/xmlns/", "xmlns", "http://www.w3.org/2000/svg");
          
    svg.node().setAttributeNS("http://www.w3.org/2000/xmlns/", "xmlns:xlink", "http://www.w3.org/1999/xlink");

In case it's helpful for future search engine visitors, here's the [stackoverflow question](http://stackoverflow.com/questions/18092976/is-it-possible-to-import-svg-shapes-in-d3-js) that led me here, along with a  [bl.ock](http://bl.ocks.org/explunit/5988971) that shows an example of these attributes hard-coded in an `<svg>` tag.

**Comment by mbostock** [comment]
Seems like XML is a special namespace, too? But I can’t think of any examples where we use an "xml:…" attribute.

**Comment by mbostock** [comment]
Hmm. Seems like only XMLNS should be treated specially, according to empirical testing. So maybe we’ll start there.

**Comment by mbostock** [comment]
Fixed in #2591 for 3.5.7.

**Comment by micahstubbs** [comment]
:+1:
